serial-settings/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+                item_type: menu::ItemType::Callback {
+                    function: Self::handle_list,
+                    parameters: &[],
+                },
+            },
+            &menu::Item {
+                command: "get",
+                help: Some("Read a setting_from the device."),
+                item_type: menu::ItemType::Callback {
+                    function: Self::handle_get,
+                    parameters: &[menu::Parameter::Mandatory {
+                        parameter_name: "item",
+                        help: Some("The name of the setting to read."),
+                    }]
+                },
+            },
+            &menu::Item {
+                command: "set",
+                help: Some("Update a a setting in the device."),
+                item_type: menu::ItemType::Callback {
+                    function: Self::handle_set,
+                    parameters: &[
+                        menu::Parameter::Mandatory {
+                            parameter_name: "item",
+                            help: Some("The name of the setting to write."),
+                        },
+                        menu::Parameter::Mandatory {
+                            parameter_name: "value",
+                            help: Some("Specifies the value to be written. Values must be JSON-encoded"),
+                        },
+                    ]
+                },
+            },
+            &menu::Item {
+                command: "clear",
+                help: Some("Clear the device settings to default values."),
+                item_type: menu::ItemType::Callback {
+                    function: Self::handle_clear,
+                    parameters: &[]
+                },
+            },
+            &menu::Item {
+                command: "platform",
+                help: Some("Platform specific commands"),
+                item_type: menu::ItemType::Callback {
+                    function: Self::handle_platform,
+                    parameters: &[menu::Parameter::Mandatory {
+                        parameter_name: "cmd",
+                        help: Some("The name of the command (e.g. `reboot`, `service`, `dfu`)."),
+                    }]
+                },
+            },
+        ],
+        entry: None,
+        exit: None,
+    }
+    }
+}
+
+impl<'a, P: Platform> core::fmt::Write for Context<'a, P> {
+    /// Write data to the serial terminal.
+    ///
+    /// # Note
+    /// The terminal uses an internal buffer. Overflows of the output buffer are silently ignored.
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
+        let mut writer = Writer(self.platform.interface_mut());
+        writer.write_str(s)
+    }
+}
+
+struct Writer<'a, T: WriteReady + EioWrite>(&'a mut T);
+
+impl<'a, T: embedded_io::WriteReady + embedded_io::Write> core::fmt::Write
+    for Writer<'a, T>
+{
+    fn write_str(&mut self, s: &str) -> core::fmt::Result {
```

Comment:
I still have an issue with this. Why does this crate need to make the write best effort? That seems like a thing for a downstream crate to decide.

Reply:
Good point. I'll update it to not make assumptions about the writability of data.

Reply:
I suspect that will also allow removing the dependency on eio traits from this crate.

Reply:
We need the `Read` component as well for the `process()` function to feed data to the menu interface. While we could replace the eio Write methods, it feels really odd to use half e-io and half core::fmt

Reply:
agreed

Reply:
But this does tie into the notion that `Runner` doesn't do much other than encapsulating `eio::Read`.

Reply:
Yeah, I recognize that. It largely comes down to whether we want this to be a "batteries included" approach or not. While I'm generally not a huge fan of making assumptions for the user, I think the `Runner` does eliminate a fair amount of boilerplate for the end user. Additionally, I added blanket-impls of the e-io traits for `usbd_serial::SerialPort`, so USB serial devices won't need any additional code to handle reading/writing data.

Ultimately, I'd say that we should keep the status quo and re-evaluate in the future if we want.